ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,543 @@
+    while let Some(event) = observer.next().await {
+        match event? {
+            ObserverEvent::Blob(_blob) => {
+                unreachable!("no blob events should occur before chain head is synced")
+            }
+            ObserverEvent::Block(_) => {}
+            ObserverEvent::BlockSynced(data) => {
+                debug_assert_eq!(highest_block, data.block_hash);
+                break;
+            }
+        }
+    }
+
+    Ok(highest_block)
+}
+
+async fn sync_from_network(
+    network: &mut NetworkService,
+    db: &Database,
+    program_states: &BTreeMap<ActorId, H256>,
+) {
+    let mut manager = RequestManager::default();
+    for &state in program_states.values() {
+        manager.add(state, RequestMetadata::ProgramState);
+    }
+
+    loop {
+        let (completed, pending) = manager.stats();
+        log::info!("[{completed:>05} / {pending:>05}] Getting network data");
+
+        let Some(responses) = manager.request(network, db).await else {
+            break;
+        };
+
+        for (metadata, data) in responses {
+            match metadata {
+                RequestMetadata::ProgramState => {
+                    let state: ProgramState =
+                        Decode::decode(&mut &data[..]).expect("`db-sync` must validate data");
+
+                    let ProgramState {
+                        program,
+                        queue_hash,
+                        waitlist_hash,
+                        stash_hash,
+                        mailbox_hash,
+                        balance: _,
+                        executable_balance: _,
+                    } = &state;
+
+                    if let Program::Active(ActiveProgram {
+                        allocations_hash,
+                        pages_hash,
+                        memory_infix: _,
+                        initialized: _,
+                    }) = program
+                    {
+                        if let Some(allocations_hash) = allocations_hash.hash() {
+                            manager.add(allocations_hash, RequestMetadata::Data);
+                        }
+                        if let Some(pages_hash) = pages_hash.hash() {
+                            manager.add(pages_hash, RequestMetadata::MemoryPages);
+                        }
+                    }
+
+                    if let Some(queue_hash) = queue_hash.hash() {
+                        manager.add(queue_hash, RequestMetadata::Data);
+                    }
+                    if let Some(waitlist_hash) = waitlist_hash.hash() {
+                        manager.add(waitlist_hash, RequestMetadata::Data);
+                    }
+                    if let Some(mailbox_hash) = mailbox_hash.hash() {
+                        manager.add(mailbox_hash, RequestMetadata::Mailbox);
+                    }
+                    if let Some(stash_hash) = stash_hash.hash() {
+                        manager.add(stash_hash, RequestMetadata::Data);
+                    }
```

Comment:
IMPORTANT 

Maybe I miss something, but all this structures contains payloads inside. Payloads could be provided by Hash, which is also need to be loaded, but as I see you use Data for most of them and also below where UserMailbox requested

Reply:
great catch